File: Application/src/main/java/com/shareyourproxy/api/rx/command/AddUserChannelCommand.java
package com.shareyourproxy.api.rx.command;

import android.app.Service;
import android.os.Parcel;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.shareyourproxy.api.domain.model.Channel;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.rx.RxBusDriver;
import com.shareyourproxy.api.rx.RxUserChannelSync;
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback;

import java.util.List;

/**
 * Created by Evan on 6/8/15.
 */
public class AddUserChannelCommand extends BaseCommand {
    public static final Creator<AddUserChannelCommand> CREATOR = new Creator<AddUserChannelCommand>
        () {
        @Override
        public AddUserChannelCommand createFromParcel(Parcel in) {
            return new AddUserChannelCommand(in);
        }

        @Override
        public AddUserChannelCommand[] newArray(int size) {
            return new AddUserChannelCommand[size];
        }
    };
    private final static java.lang.ClassLoader CL = AddUserChannelCommand.class.getClassLoader();
    public final User user;
    public final Channel newChannel;
    public final Channel oldChannel;

    public AddUserChannelCommand(
        @NonNull RxBusDriver rxBus, @NonNull User user, @NonNull Channel newChannel,
        @Nullable Channel oldChannel) {
        super(AddUserChannelCommand.class.getPackage().getName(),
            AddUserChannelCommand.class.getName(), rxBus);
        this.user = user;
        this.newChannel = newChannel;
        this.oldChannel = oldChannel;
    }

    public AddUserChannelCommand(
        @NonNull RxBusDriver rxBus, @NonNull User user, @NonNull Channel newChannel) {
        super(AddUserChannelCommand.class.getPackage().getName(),
            AddUserChannelCommand.class.getName(), rxBus);
        this.user = user;
        this.newChannel = newChannel;
        this.oldChannel = null;
    }

    private AddUserChannelCommand(Parcel in) {
        this((RxBusDriver) in.readValue(CL),
            (User) in.readValue(CL), (Channel) in.readValue(CL), (Channel) in.readValue(CL));
    }

    @Override
    public List<EventCallback> execute(Service service) {
        return RxUserChannelSync.saveUserChannel(
            service, rxBus, user, oldChannel, newChannel);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeValue(user);
        dest.writeValue(newChannel);
        dest.writeValue(oldChannel);
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/command/AddUserChannelCommand.kt
package com.shareyourproxy.api.rx.command

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.shareyourproxy.api.domain.model.Channel
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.api.rx.RxUserChannelSync.saveUserChannel
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback

/**
 * Add a channel.
 */
internal final class AddUserChannelCommand(val user: User, val newChannel: Channel, val oldChannel: Channel) : BaseCommand() {

    constructor(user: User, newChannel: Channel) : this(user, newChannel, Channel())

    private constructor(parcel: Parcel) : this(parcel.readValue(CL) as User, parcel.readValue(CL) as Channel, parcel.readValue(CL) as Channel)

    override fun execute(context: Context): EventCallback {
        return saveUserChannel(context, user, oldChannel, newChannel)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(user)
        dest.writeValue(newChannel)
        dest.writeValue(oldChannel)
    }

    companion object {
        private val CL = AddUserChannelCommand::class.java.classLoader
        @JvmField val CREATOR: Parcelable.Creator<AddUserChannelCommand> = object : Parcelable.Creator<AddUserChannelCommand> {
            override fun createFromParcel(parcel: Parcel): AddUserChannelCommand {
                return AddUserChannelCommand(parcel)
            }

            override fun newArray(size: Int): Array<AddUserChannelCommand?> {
                return arrayOfNulls(size)
            }
        }
    }
}
